web/modules/custom/ct_user/ct_user.module

```diff
@@ -22,3 +23,47 @@ function ct_user_help($route_name, RouteMatchInterface $route_match) {
     default:
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function ct_user_form_alter(&$form, $form_state, $form_id) {
+  // Check if the form is the user login form.
+  if ($form_id == 'user_login_form') {
+    // Add a custom validation function to the form.
+    $form['#validate'][] = 'ct_user_user_login_form_validate';
+  }
+}
+
+/**
+ * Custom validation function for the user login form.
+ */
+function ct_user_user_login_form_validate(&$form, $form_state) {
+  // Get the name value from the form.
+  $name = $form_state->getValue('name');
+
+  // Get the id of the user.
+  $ids = \Drupal::entityQuery('user')
+    ->condition('name', $name)
+    ->range(0, 1)
+    ->accessCheck()
+    ->execute();
+
+  if ($ids) {
+    // Check if there is a social auth profile.
+    $socialIds = \Drupal::entityQuery("social_auth")
+      ->accessCheck(FALSE)
+      ->condition("user_id", reset($ids))
+      ->execute();
+  }
+
+  // If the user has no social auth profile, allow the login to continue.
+  if (empty($socialIds)) {
+    return;
+  }
+
+  // If the user has social auth profile, redirect them to /user/login/google.
+  $response = new RedirectResponse('/user/login/google');
+  $response->send();
+  user_logout();
+}
```

Comment:
Consider the flow:

If no user with that email exists, `$ids` will be empty. In that case, the first `if` block never gets executed and `$socialIds` is never set. This means that the second block won't be entered. Finally, in this case, the user will get redirected to Google login. This might be okay in registration case but it is not very likely that someone will come this way to register. It will be clearer to the user if they are actually sent back. In our case, we just have to return from validation because the login system in Drupal will throw the error.

The root of this problem here is that the second `if` block is outside of the first `if` block and there is a chance it never gets set. You won't see errors because you are checking for `empty` but the problem still remains (as I have described above).

To fix this, you can make it very simple by considering this pseudo-code.
1. If the user doesn't exists (`$ids` is empty), then immediately return.
2. Now we know that the user definitely exists. Check if there is a social auth profile. If not, the immediately exit.
3. Now we know that the user exists AND has a social auth profile. Redirect to Google auth.

Reply:
Hi, I have updated the code. Please verify